### PR TITLE
Added Memory Appender, Log assertions in OpSparkListner

### PR DESCRIPTION
**Pull request by ajayborra**
- Added Memory Appender, Log assertions for OpSparkListner

### CODE REVIEW

File: utils/src/main/scala/com/salesforce/op/test/MemoryAppender.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ */
+
+package com.salesforce.op.test
+
+import org.apache.log4j.{AppenderSkeleton, Level, spi}
+
+import scala.collection.mutable
+
+class MemoryAppender extends AppenderSkeleton {
+  private val _logs = new mutable.HashSet[spi.LoggingEvent]
```
Comment (tovbinm) [inline utils/src/main/scala/com/salesforce/op/test/MemoryAppender.scala]: 1. no need to prefix with `_`
2. make this an `ArrayBuffer`

File: utils/src/main/scala/com/salesforce/op/test/MemoryAppender.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
+/*
```
Comment (tovbinm) [inline utils/src/main/scala/com/salesforce/op/test/MemoryAppender.scala]: define this appender in the same file as `OpSparkListenerTest` - we dont need to expose it as it's currently only useful in one test.

File: utils/src/main/scala/com/salesforce/op/test/TestSparkContext.scala [hunk 3]
```diff
@@ -44,6 +44,15 @@ trait TestSparkContext extends TempDirectoryTest with TestCommon {
     classOf[com.salesforce.op.test.PassengerCSV]
   )
 
+  val sparkLogAppender: MemoryAppender = {
```
Comment (tovbinm) [inline utils/src/main/scala/com/salesforce/op/test/TestSparkContext.scala]: we dont need this in every test. let's only use it in OpSparkListenerTest

File: utils/src/test/resources/log4j.properties [hunk 4]
```diff
@@ -12,3 +12,4 @@ log4j.logger.org.eclipse.jetty=ERROR
 
 # TransmogrifAI logging
 log4j.logger.com.salesforce.op=ERROR
+log4j.logger.com.salesforce.op.utils.spark.OpSparkListener=INFO
```
Comment (tovbinm) [inline utils/src/test/resources/log4j.properties]: you can set this `log4j.logger.com.salesforce.op.utils.spark.OpSparkListener=INFO` programmatically in OpSparkListenerTest

File: utils/src/main/scala/com/salesforce/op/test/MemoryAppender.scala [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+   */
+  def logs: mutable.HashSet[spi.LoggingEvent] = _logs
+
+  /**
+   * Check if a log exists in the log collection
+   * @param logLevel The log level of the message
+   * @param logMessage The log message
+   * @return Boolean of log existence
+   */
+  def logExists(logLevel: Level, logMessage: String): Boolean = {
```
Comment (tovbinm) [inline utils/src/main/scala/com/salesforce/op/test/MemoryAppender.scala]: this seems a bit redundant. you can simply do this check in the test itself.

File: utils/src/main/scala/com/salesforce/op/test/MemoryAppender.scala (hunk 1 above)
Comment (ajayborra) [reply utils/src/main/scala/com/salesforce/op/test/MemoryAppender.scala]: Was using hash sets for constant inset and lookup times. Array buf's seems to be a good choice as well, will change it.

Comment (tovbinm) [reply utils/src/main/scala/com/salesforce/op/test/MemoryAppender.scala]: Hashset would dedup and loose order, thereforce we cannot use it.

Comment (ajayborra) [reply utils/src/main/scala/com/salesforce/op/test/MemoryAppender.scala]: Good point.

**Comment by ajayborra** [comment]
Closing this PR in favor of https://github.com/salesforce/TransmogrifAI/pull/69. As the link b/w this PR and the branch on old fork broke, post making repo public.

**Pull request closed without merging**